File: sdk/rust/inferlet/src/tools.rs
//! Optional helpers for tool calling.
//!
//! `inferlet` does not bake a tool-call loop into the SDK
//! surface — the right loop shape varies a lot between agents (ReAct,
//! CodeAct, JSON-call, native-grammar) and we'd rather give you the
//! pieces than a framework. This module exposes the host's tool-template
//! capability so callers that *do* want the model's native format can
//! reach for it explicitly:
//!
//! - [`equip_prefix`] — token sequence that registers tool schemas in
//!   the chat template (model-specific). Append to your context buffer.
//! - [`answer_prefix`] — token sequence that frames a tool result for
//!   the next turn.
//! - [`native_grammar`] — the model's tool-call grammar, if it has one.
//!   Wrap with [`GrammarConstraint`](crate::GrammarConstraint) and apply
//!   its mask each decode step to enforce well-formed output.
//! - [`Decoder`] — streaming detector for tool calls inside generated
//!   text. Feed each step's tokens; collect `Call(name, args)` events.
//! - [`parse_call`] — best-effort one-shot parse out of a finished text
//!   blob. Useful for `collect_text()` flows that want to extract one
//!   call at the end.
//!
//! For agents that hand-roll their own format (e.g. `agent-react`'s
//! `Action: ToolName[input]` parsing), none of these are required.

use crate::Result;
use crate::inference::Grammar;
use crate::pie::core::inference::Matcher;
use crate::pie::instruct::tool_use;

// =============================================================================
// Tool trait — schema metadata for chat-template registration
// =============================================================================

/// Tool metadata used by [`equip_prefix`] to splice
/// schemas into the model's chat template. Implement directly for dynamic
/// tools, or derive via the [`#[tool]`](inferlet_macros::tool) macro for
/// static ones.
///
/// `schema` returns the JSON Schema for the args **object only** —
/// `{"type":"object","properties":{...},"required":[...]}` — without the
/// outer `{name, description, parameters}` envelope. The SDK wraps it in
/// whatever shape the host's `equip_prefix` expects.
pub trait Tool {
    fn name(&self) -> &'static str;
    fn description(&self) -> &'static str;
    fn schema(&self) -> &'static str;
}

// =============================================================================
// Templating
// =============================================================================

/// Token sequence that registers `tool_schemas` (each a JSON Schema
/// string) in the chat template. Append to your prompt token buffer
/// before the user
/// message. Models without a native tool-template return an empty vec.
pub fn equip_prefix(tool_schemas: &[String]) -> Result<Vec<u32>> {
    tool_use::equip(tool_schemas)
}

/// Token sequence that registers `tools` in the chat template. Wraps each
/// tool's [`Tool::schema`] in the `{name, description, parameters}` envelope
/// the host expects, then lowers via [`equip_prefix`]. Append the result to
/// your prompt token buffer before the user message. This is the raw-WIT
/// tool-registration primitive (the analog of chat templating) — inferlets
/// call it directly instead of going through a context facade.
pub fn equip(tools: &[&dyn Tool]) -> Result<Vec<u32>> {
    let envelopes: Vec<String> = tools
        .iter()
        .map(|t| {
            let parsed: serde_json::Value = serde_json::from_str(t.schema())
                .map_err(|e| format!("tool `{}`: invalid schema: {e}", t.name()))?;
            Ok(serde_json::json!({
                "name": t.name(),
                "description": t.description(),
                "parameters": parsed,
            })
            .to_string())
        })
        .collect::<Result<_>>()?;
    equip_prefix(&envelopes)
}

/// Token sequence that frames a tool result for the next turn. `name`
/// matches the call the model made; `value` is typically a JSON-encoded
/// result.
pub fn answer_prefix(name: &str, value: &str) -> Vec<u32> {
    tool_use::answer(name, value)
}

// =============================================================================
// Native grammar / matcher
// =============================================================================

/// The model's native tool-call grammar, if any. Returns `None` when the
/// model has no enforceable format (the caller should fall through to
/// free-form generation + their own parser).
pub fn native_grammar(tool_schemas: &[String]) -> Option<Grammar> {
    tool_use::format(tool_schemas)
}

/// Build a [`Matcher`] for the model's native tool-call grammar.
/// Returns `None` when the model has no enforceable format. Pair with
/// [`GrammarConstraint::new`](crate::GrammarConstraint::new) for
/// constrained generation.
pub fn native_matcher(tool_schemas: &[String]) -> Option<Matcher> {
    Some(tool_use::create_matcher(tool_schemas))
}

// =============================================================================
// Streaming decoder
// =============================================================================

/// Streaming tool-call detector. Feed each step's accepted tokens; the
/// decoder emits [`Event::Start`] when a tool call begins and
/// [`Event::Call`] (with parsed `name` and `arguments`) when it
/// completes.
pub struct Decoder {
    inner: tool_use::Decoder,
}

/// Tool-decoding event. One per `feed`.
///
/// `Start` fires while a tool-call structure is being assembled but the
/// arguments haven't closed yet — it's both "boundary entered" and the
/// no-meaningful-event signal during accumulation. Most callers can
/// ignore it and only act on `Call`.
#[derive(Clone, Debug)]
pub enum Event {
    /// A tool call is in progress — keep feeding.
    Start,
    /// Complete tool call: `(name, arguments_json)`.
    Call(String, String),
}

impl Decoder {
    /// Construct a decoder for the bound model's tool-call template.
    pub fn new() -> Self {
        Self {
            inner: tool_use::create_decoder(),
        }
    }

    /// Feed a token batch and get back the event that fired (one per
    /// call). `Start` indicates an in-progress tool call; `Call` fires
    /// once when the arguments close.
    pub fn feed(&mut self, tokens: &[u32]) -> Result<Event> {
        match self.inner.feed(tokens)? {
            tool_use::Event::Start => Ok(Event::Start),
            tool_use::Event::Call((name, args)) => Ok(Event::Call(name, args)),
        }
    }

    /// Reset to initial state.
    pub fn reset(&mut self) {
        self.inner.reset();
    }
}

// =============================================================================
// One-shot parse
// =============================================================================

/// Best-effort parse of a single tool call out of a finished text blob.
/// Internally tokenizes, runs the decoder, and returns the first
/// completed `Call`. Useful for `collect_text()` flows that want to
/// extract one call at the end of generation. Returns `None` when no
/// completed call is detected.
pub fn parse_call(text: &str) -> Option<(String, String)> {
    let tokens = crate::model::encode(text);
    let mut dec = Decoder::new();
    match dec.feed(&tokens).ok()? {
        Event::Call(name, args) => Some((name, args)),
        Event::Start => None,
    }
}

File: sdk/inferlet/javascript/src/tools.ts
// Tool calling — `pie:inferlet/tools`.
//
// Equip tool specs into the prompt, constrain generation to well-formed
// calls, and parse the calls back out.

import * as _tools from 'pie:inferlet/tools@0.3.0';

import { Grammar, GrammarError, Matcher } from './grammar.js';

export interface ToolCall {
  readonly name: string;
  readonly argumentsJson: string;
}

export interface EventStart {
  readonly type: 'start';
}

export interface EventCall {
  readonly type: 'call';
  readonly call: ToolCall;
}

export type Event = EventStart | EventCall;

/** Token sequence declaring `tools` (JSON specs) to the model. */
export function equip(tools: string[]): Uint32Array {
  try {
    return _tools.equip(tools);
  } catch (e: unknown) {
    const payload = (e as { payload?: unknown })?.payload;
    throw new GrammarError(`tools.equip: ${typeof payload === 'string' ? payload : String(e)}`);
  }
}

/** Token sequence returning a tool's result to the model. */
export function answer(name: string, value: string): Uint32Array {
  return _tools.answer(name, value);
}

/** The grammar of a well-formed call to one of `tools`, if the template has one. */
export function format(tools: string[]): Grammar | undefined {
  const g = _tools.format(tools);
  return g ? new Grammar(g) : undefined;
}

/** A matcher over `format`'s grammar. */
export function createMatcher(tools: string[]): Matcher {
  return new Matcher(_tools.createMatcher(tools));
}

/** Parses generated tokens into tool-call events. */
export class Decoder {
  readonly #inner = new _tools.Decoder();

  feed(tokens: ArrayLike<number>): Event {
    let ev: _tools.Event;
    try {
      ev = this.#inner.feed(Uint32Array.from(tokens));
    } catch (e: unknown) {
      const payload = (e as { payload?: unknown })?.payload;
      throw new GrammarError(`tools decoder: ${typeof payload === 'string' ? payload : String(e)}`);
    }
    if (ev.tag === 'call') return { type: 'call', call: { name: ev.val.name, argumentsJson: ev.val.argumentsJson } };
    return { type: 'start' };
  }

  reset(): void {
    this.#inner.reset();
  }
}
